recovery/elliptics_recovery/dc_server_send.py

```diff
@@ -0,0 +1,206 @@
+
+import elliptics
+
+log = logging.getLogger()
+
+
+class BucketsManager(object):
+    def __init__(self, ctx):
+        self.ctx = ctx
+        self.bucket_index = -1
+        self.buckets = dict()
+
+    def get_next_bucket(self):
+        self.bucket_index = (self.bucket_index + 1) % len(self.ctx.bucket_order)
+        group = self.ctx.bucket_order[self.bucket_index]
+        log.debug("Get next bucket: index: {0}, group: {1}, bucket_order: {2}".format(self.bucket_index, group, self.ctx.bucket_order))
+        return self._get_bucket(group)
+
+    def on_server_send_fail(self, key, key_infos, next_group):
+        if next_group >= 0:
+            b = self._get_bucket(next_group)
+            b.add_key(key, key_infos)
+        else:
+            self.move_to_rest_bucket(key, key_infos)
+
+    def move_to_rest_bucket(self, key, key_infos):
+        log.debug("Moving key to rest keys bucket: {0}".format(key))
+        key_data = (key, key_infos)
+        dump_key_data(key_data, self.ctx.rest_file)
+
+    def _get_bucket(self, group):
+        if group not in self.buckets:
+            if group not in self.ctx.bucket_files:
+                filename = os.path.join(self.ctx.tmp_dir, 'bucket_%d' % (group))
+                self.ctx.bucket_files[group] = open(filename, 'wb+')
+                self.ctx.bucket_order.append(group)
+            self.buckets[group] = BucketKeys(self.ctx.bucket_files[group], group)
+        return self.buckets[group]
+
+
+class BucketKeys(object):
+    def __init__(self, bucket_file, group):
+        log.debug("Create bucket: group: {0}, bucket: {1}".format(group, bucket_file.name))
+        self.bucket_file = bucket_file
+        self.group = group
+
+    def get_keys(self, max_keys_num):
+        self.bucket_file.seek(0)
+        keys = []
+        for key, key_infos in load_key_data_from_file(self.bucket_file):
+            keys.append((key, key_infos))
+            if len(keys) >= max_keys_num:
+                yield keys
+                keys = []
+        if keys:
+            yield keys
+
+    def get_group(self):
+        return self.group
+
+    def clear(self):
+        log.debug("Clear bucket: group: {0}, bucket: {1}".format(self.group, self.bucket_file.name))
+        self.bucket_file.seek(0)
+        self.bucket_file.truncate()
+
+    def add_key(self, key, key_infos):
+        log.debug("Append key to bucket: group: {0}, bucket: {1}".format(self.group, self.bucket_file.name))
+        key_data = (key, key_infos)
+        dump_key_data(key_data, self.bucket_file)
+
+
+class ServerSendRecovery(object):
+    def __init__(self, ctx, node):
+        self.ctx = ctx
+        self.stats = ctx.stats['recover']
+        self.node = node
+        self.buckets = BucketsManager(ctx)
+
+        self.session = elliptics.Session(node)
+        self.session.trace_id = ctx.trace_id
+        self.session.exceptions_policy = elliptics.exceptions_policy.no_exceptions
+        self.session.timeout = 120 # magic ?
```

Comment:
Why is it magic?

120 seconds should probably be large enough to wait for the first iterator result since it may skip many keys, but in server-send case results should be received very quickly after beginning of the call